src/tools/fuzzing/heap-types.cpp

```diff
@@ -0,0 +1,499 @@
+    if (auto type = pickKind<DataKind>()) {
+      return *type;
+    } else {
+      return HeapType::data;
+    }
+  }
+
+  HeapType pickSubEq() {
+    if (rand.oneIn(2)) {
+      return HeapType::i31;
+    } else {
+      return pickSubData();
+    }
+  }
+
+  HeapType pickSubAny() {
+    switch (rand.upTo(4)) {
+      case 0:
+        return HeapType::func;
+      case 1:
+        return HeapType::eq;
+      case 2:
+        return pickSubFunc();
+      case 3:
+        return pickSubEq();
+    }
+    WASM_UNREACHABLE("unexpected index");
+  }
+
+  HeapType pickSubHeapType(HeapType type) {
+    auto it = typeIndices.find(type);
+    if (it != typeIndices.end()) {
+      // This is a constructed type, so we know where its subtypes are.
+      return builder[rand.pick(subtypeIndices[typeIndices[type]])];
+    } else {
+      // This is not a constructed type, so it must be a basic type.
+      assert(type.isBasic());
+      switch (type.getBasic()) {
+        case HeapType::func:
+          return pickSubFunc();
+        case HeapType::ext:
+          return HeapType::ext;
+        case HeapType::any:
+          return pickSubAny();
+        case HeapType::eq:
+          return pickSubEq();
+        case HeapType::i31:
+          return HeapType::i31;
+        case HeapType::data:
+          return pickSubData();
+      }
+      WASM_UNREACHABLE("unexpected kind");
+    }
+  }
+
+  // TODO: Make this part of the wasm-type.h API
+  struct Ref {
+    HeapType type;
+    Nullability nullability;
+  };
+
+  Ref generateSubRef(Ref super) {
+    auto nullability = super.nullability == NonNullable
+                         ? NonNullable
+                         : rand.oneIn(2) ? Nullable : NonNullable;
+    return {pickSubHeapType(super.type), nullability};
+  }
+
+  Rtt generateSubRtt(Rtt super) {
+    auto depth = super.hasDepth()
+                   ? super.depth
+                   : rand.oneIn(2) ? Rtt::NoDepth : rand.upTo(MAX_RTT_DEPTH);
+    return {depth, super.heapType};
+  }
+
+  Type generateSubtype(Type type) {
+    if (type.isBasic()) {
+      // We do not construct types with basic reference types (we go through the
+      // TypeBuilder for those instead), so this must be a non-reference basic
+      // type, which means it has no other subtypes.
+      return type;
+    } else if (type.isRef()) {
+      auto ref = generateSubRef({type.getHeapType(), type.getNullability()});
+      return builder.getTempRefType(ref.type, ref.nullability);
+    } else if (type.isRtt()) {
+      auto rtt = generateSubRtt(type.getRtt());
+      return builder.getTempRttType(rtt);
+    } else {
+      WASM_UNREACHABLE("unexpected type kind");
+    }
+  }
+
+  Signature generateSubSignature(Signature super) {
+    // TODO: Update this once we support nontrivial function subtyping.
+    return super;
+  }
+
+  Field generateSubField(Field super) {
+    if (super.mutable_ == Mutable) {
+      // Only immutable fields support subtyping.
```

Comment:
This has been the case all along because mutable subtyping would be unsound. For example, if `(struct (mut funcref))` were a subtype of `(struct (mut anyref))`, then I could write a non-func value via the `(mut anyref)`, then read it as a funcref from the `(mut funcref)`.

Reply:
I don't get it... The subtyping is `funcref`<:`anyref`, then how can we write a non-func value into `mut anyref` and read it as a `funcref` from `mut funcref`?

Reply:
Here's an example:

```
(module
  (type $func (func))
  (type $super (array (mut anyref)))
  (type $sub (array (mut (ref null $func)))) ;; subtype of $super
  (import "some" "extern" (global $extern externref))
  
  ;; Write an `externref` at index 0 of a `$super` value.
  (func $write-extern (param $x (ref $super))
    (array.set
      (local.get $x)
      (i32.const 0)
      (global.get $extern)
    )
  )

  ;; Call $write-extern, then read a (ref null func) out of the $sub value and call it.
  (func $unsound (param $x (ref $sub))
    (call $write-extern
      (local.get $sub)
    )
    (call_ref
      (array.get
        (local.get $sub)
        (i32.const 0)
      )
    )
  )
)
```